File: salyangoz.cs
using System;

class Program
{
    static int[,] salyangozMatrisiOlustur(int boyut)
    {
        //salyangoz matrisini tutmak için 2 boyutlu int dizisi oluşturuyoruz.
        int[,] matris = new int[boyut, boyut];

        //Alttaki Initialize() fonksiyonu ile dizinin elemanlarını varsayılan değerlerine yani sıfıra eşitliyoruz.
        //Bir elemanın sıfır olması aynı zamanda henüz o elemanın atanmadığı anlamına gelecektir.
        //Yani aşağıdaki kodlarda bu matrisin elemanının sıfıra eşit olup olmamasına göre işlem yapacağız.
        matris.Initialize();

        //Burada matris içinde salyangoz çizip gerekli değerleri atamak için gereken değişkenler mevcut.
        int apsisGezgini = 0, ordinatGezgini = 0, toplam = 1;

        //Bu değişken true olduğunda apsis ve ordinat gezginlerini arttırırız aksi halde azaltırız.
        bool arttir = true;

        //Burada boyut sayısı kadar döngü tekrar ettiğinde salyangoz tamamlanıyor.
        for (int i = 0; i < boyut; i++)
        {
            //apsis ve ordinat gezginlerinin arttırılmadan önceki değerleri atanıyor.
            matris[apsisGezgini, ordinatGezgini] = toplam;

            if (arttir) //Bu alanda hem apsis hem ordinat gezginlerinin değerleri max alabilecekleri kadar arttırılıyor.
            {
                //Arttırma işlemi yapmak için bir sonraki koordinatın dolu olup olmadığına bakıyoruz.
                while (apsisGezgini + 1 < boyut && matris[apsisGezgini + 1, ordinatGezgini] == 0)
                {
                    apsisGezgini++; //apsisGezginini 1 arttırıyoruz.
                    toplam++;       //toplam değişkenini 1 arttırıyoruz.

                    //Matrisimizde şuandaki noktanın değeri 'toplam' değişkenine eşit olduğu için 
                    //gerekli atamayı yapıyoruz.
                    matris[apsisGezgini, ordinatGezgini] = toplam;
                }
                //Bir üstteki işlemin aynısı ordinat gezgini için yapılıyor.
                while (ordinatGezgini + 1 < boyut && matris[apsisGezgini, ordinatGezgini + 1] == 0)
                {
                    ordinatGezgini++;
                    toplam++;
                    matris[apsisGezgini, ordinatGezgini] = toplam;
                }
            }
            else //Bu alanda hem apsis hem ordinat gezginlerinin değerleri minimum alabilecekleri kadar azaltıyoruz.
            {
                //Azaltma işlemi yapmak için bir önceki koordinatın dolu olup olmadığına bakıyoruz.
                while (apsisGezgini > 0 && matris[apsisGezgini - 1, ordinatGezgini] == 0)
                {
                    apsisGezgini--; //apsisGezginini 1 azaltıyoruz.
                    toplam++;       //toplam değişkenini 1 arttırıyoruz.

                    //Matrisimizde şuandaki noktanın değeri 'toplam' değişkenine eşit olduğu için 
                    //gerekli atamayı yapıyoruz.
                    matris[apsisGezgini, ordinatGezgini] = toplam;
                }
                //Bir üstteki işlemin aynısı ordinat gezgini için yapılıyor.
                while (ordinatGezgini > 0 && matris[apsisGezgini, ordinatGezgini - 1] == 0)
                {
                    ordinatGezgini--;
                    toplam++;
                    matris[apsisGezgini, ordinatGezgini] = toplam;
                }
            }
            //arttir değişkeninin tersini alıyoruz ki üstteki if yapısının blokları sırasıyla işlesin.
            arttir = !arttir;
        }
        //Oluşturulan salyangoz matrisini döndürüyoruz.
        return matris;
    }

    static void Main(string[] args)
    {
        //Matris boyutunu tutan değişken.
        int matrisBoyutu;

        //Kullanıcıdan matris boyutu girmesini istiyoruz.
        Console.WriteLine("Salyangoz matrisi için boyut giriniz:");

        //Girilen matris boyutunu alıyoruz.
        matrisBoyutu = int.Parse(Console.ReadLine());

        //Üstte yazdığımız 2 boyutlu int türünde dizi döndüren 'salyangozMatrisiOlustur' ismindeki 
        //fonksiyonu çağırarak istenilen boyutta salyangoz matrisi oluşturuyoruz.
        int[,] matris = salyangozMatrisiOlustur(matrisBoyutu);

        //Oluşturduğumuz salyangoz matrisini ekrana yazdırıyoruz.
        for (int satir = 0; satir < matrisBoyutu; satir++)
        {
            for (int sutun = 0; sutun < matrisBoyutu; sutun++)
            {
                //Burada ekrana yazılan her değerin yanına düzenli görünmesi için tab boşluğu ekleniyor.
                Console.Write(matris[sutun, satir] + "\t");
            }
            //Alt satıra geçiyoruz.
            Console.WriteLine();
        }
        Console.ReadKey();
    }
}

File: salyangoz.py
n = int(input("Kare matrisin n sayısını giriniz: "))

matris = [[0 for x in range(n)] for y in range(n)]

ardisik_sayilar = [x for x in range(1, (n**2)+1)]

satir_basi = 0
satir_sonu = len(matris)
sutun_basi = 0
sutun_sonu = len(matris[0])

sayac = 0

while satir_basi < satir_sonu and sutun_basi < sutun_sonu:
    # Satırın soldan sağa yazdırılması
    for j in range(sutun_basi, sutun_sonu):
        matris[sutun_basi][j] = ardisik_sayilar[sayac]
        sayac += 1
    
    satir_basi += 1
    
    # Sütunun yukarıdan aşağıya yazdırılması
    for i in range(satir_basi, satir_sonu):
        matris[i][sutun_sonu-1] = ardisik_sayilar[sayac]
        sayac += 1
    
    sutun_sonu -= 1
    
    # Satırın sağdan sola yazdırılması
    if satir_basi < satir_sonu:
        for j in range(sutun_sonu-1, sutun_basi-1, -1):
            matris[satir_sonu-1][j] = ardisik_sayilar[sayac]
            sayac += 1
    
    satir_sonu -= 1
    
    # Sütunun aşağıdan yukarıya yazdırılması
    if sutun_basi < sutun_sonu:
        for i in range(satir_sonu-1, satir_basi-1, -1):
            matris[i][sutun_basi] = ardisik_sayilar[sayac]
            sayac += 1
    
    sutun_basi += 1

for i in range(n):
    for j in range(n):
        print(matris[i][j], end="\t")
    
    print()
